ys001-yggdrasil-core-specification.md

```diff
@@ -0,0 +1,943 @@
+
+Repeat the process by right-shifting 7 bytes from the source value. If the
+shifted source value is still greater than zero, append the next byte by taking
+the next 7 least-significant bits and set the most-significant bit to 1.
+
+The resultant bytes will be big-endian individually, however the order of the
+byte array will be reversed as the beginning of the array now contains the
+source value's least-significant bits. Therefore, to complete the encoding, the
+order of the bytes must be reversed (preserving the order of the bits in each
+byte) so that the entire sequence is big-endian and the last byte in the
+sequence now has the zero most-significant bit.
+
+##### Decoding
+
+To decode a `varu64`, start with a target zero-value unsigned 64-bit integer.
+Take the 7 least-significant bits of the first byte and copy them onto the 7
+least-significant bits of the target. If the most-significant bit of the byte is
+1, left-shift the target value by 7 bits and then move onto the next byte,
+repeating this process until the most-significant bit is 0.
+
+#### Signed 64-bit integers
+
+All signed integers on the wire are encoded as `vari64`, using the same
+compression as `varu64`, specified above. However, for signed integers, zig-zag
+encoding is also used in order to preserve whether a value is positive or
+negative.
+
+Positive values of `x` are encoded as `2*x+0` and negative values of `x` are
+encoded as `2*(^x)+1`, allowing negative numbers to be complemented - in this
+case, bit 0 denotes whether to complement or not.  
+
+#### Byte arrays
+
+Byte arrays are written to the wire unmodified. There is no encoding or decoding
+to be done when working with byte arrays.
+
+As they are not prefixed with a length value, all byte arrays that occupy fields
+before the last field of the message **must** be fixed-length and that length
+must be specified and known. Byte arrays **must** only be variable-length if
+they are the last field in the message.
+
+#### Coordinates
+
+Coordinates are encoded into byte arrays, however, they have special behaviour.
+Coordinates are variable-length arrays of `varu64`s, therefore encoding
+coordinates is a three-step process:
+
+1. Encode each `varu64` element of the coordinates into a `bytes` array, as
+   above
+2. Concatenate all resultant `bytes` arrays into one contiguous `bytes` array
+3. Prefix the contiguous `bytes` array with a `varu64` which contains the
+   final array length
+
+To decode `coords`, perform the same process in reverse:
+
+1. Decode the beginning of the bytes array as a `varu64` which will show the
+   length in bytes for this set of coordinates, advancing by the number of bytes
+  used by the `varu64` length field
+2. Continue to decode the next bytes as a `varu64`, appending the result to the
+   discovered coordinates
+3. Continue through the bytes until the initial length value has been reached
+
+### Message types
+
```

Comment:
When a socket connection is initialized with a link, a `meta` message is exchanged.

Format of that message is

1. `bytes` ['m', 'e', 't', 'a']
2. `varu64` Major version
3. `varu64` Minor version
4. `bytes` Encryption key (32 bytes)
5. `bytes` Signing key (32 bytes)
6. `bytes` Encryption key for this connection (32 bytes)

Reply:
This is deliberately *not* in the message types at the moment because it's not really a true protocol message — in our current implementation it is specific to TCP/TLS peerings. I am not sure if it is best to add a section specifically about TCP/TLS, or whether that should really be an extension spec.